### ISSUE TITLE
Support for multiple rate limits

### ISSUE THREAD

**Issue by sirloon** [traefik/traefik#10274]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

I need to implement rate limits using different throttling classes. For instance, I would need:
- a "standard" rate limit of avg=100,burst=200,period=1
- a "degraded" rate limit of  avg=10,burst=20,period=60

I've tried to use multiple rateLimit middlewares stacked upon each others:
- rate-limit-standard, with the limits above, based of source criterion request header `x-throttling-class-standard`
- rate-limit-degraded, with header `x-throttling-class-degraded`

For context, these headers are populated by an upstream middleware, depending on eg. a username value, `x-throttling-class-standard: bob`, or `x-throttling-class-degraded: alice`:
1. a middleware sets a default header `x-jwt-username: __anonymous__`, so at least there's a header even if not authenticated
2. a plugin extract JWT claims, if any, and puts a username in `x-jwt-username` header, overwriting the previous value
3. I have a custom [plugin](https://github.com/sirloon/traefik-header-class-marker) which associate a user to a throttling class, eg. bob is declared in class "standard", so when bob is found in `x-jwt-username`, a header `x-throttling-class-standard: bob` is created, matching one of the rate limit middleware. There's a default class when there's no explicit match, so there's always throttling, even if not auth'd (see __anonymous__ previously set)

The code logic makes this impossible though, mostly because even if the header does not exists, a bucket for rate limits is created (basically for an empty string value), which makes all middlewares active. So as `bob`, you match the standard middleware, but also the degraded one with "" value, because there's no `x-throttling-class-degraded`.

I should have a pull request as a starting point, allowing to skip a rate limit middleware if the source criterion doesn't return a an actual value, still navigating all the moving parts and learning go..., but would also like to see if I completely overlooked something and/or if there's some interest.

Thanks!

**Comment by sirloon** [comment]
This would basically be this: https://github.com/sirloon/traefik/blob/multiple-ratelimits/pkg/middlewares/ratelimiter/rate_limiter.go#L146. If the sourceMatcher returns an empty source, the source criterion is not found, there's no match. Now there's an option to skip the rate limit if there's no match, and serve the http request (ie. next middleware).

**Comment by sirloon** [comment]
Hi @nmengin, so sorry for the late reply, I think I missed the github notifications... Thanks for getting back on this.

As for the workaround, my understanding is I couldn't really apply that in my case, because I can't have multiple routers with different rules, I need to have different rate-limits for one given router with its own rules.

I agree extending the rate-limit middleware to support multiple limits would be better, but I also thought allowing to "stack" them would be easier to implement while providing lots of flexibility. In the end, what I propose is giving an opportunity to interpret what it means when the source criterion doesn't match anything. It's really only an "if", checking if the source criterion (eg. header) has matched something or not, and if not (ie. no header), deciding if:

- an empty value should be used to create a bucket for rate-limit, this is the current behavior (and we can make this change backward compatible), which has also triggered some questions in the past, like this one: https://github.com/traefik/traefik/issues/6965
- an empty value should be interpreted as "there's no match, I will skip the rate-limit and delegate the request process to next middleware, whatever it is.

Basically, on this [line](https://github.com/traefik/traefik/blob/0ee377bc9f036124b063e7abc3f0958d51ace5fb/pkg/middlewares/ratelimiter/rate_limiter.go#L133), there's no check on whether `source` is actually holding a value, like a header value. If there's no header, `source` is an empty string and the logic continues, a bucket is always [created](https://github.com/traefik/traefik/blob/0ee377bc9f036124b063e7abc3f0958d51ace5fb/pkg/middlewares/ratelimiter/rate_limiter.go#L144), regardless of the `source` value, and rate-limit is always active even if the source criterion didn't match.

**Comment by nmengin** [comment]
Hello @sirloon,

Thank you for your feedback.

As I mentioned above, we'd love to have more feedback from the community before moving forward on your suggestion.
I propose to let it 5 extra months and check the traction before we make a decision.

**Comment by Karl-koder** [comment]
We're in the exact same situation. We need to be able to ratelimit per client but at the same time also a maximum of calls to a specific route.
Also as @sirloon mentioned, having different  throttle classes would be neat. 

Are there any news regarding this topic?

**Comment by nmengin** [comment]
Hello @sirloon,

This proposal was opened 6 months ago with almost no traction.
We won't add it to the Roadmap yet and we prefer to close it.
We'll re-open it later if necessary.

Thank you @sirloon for the suggestion and @Karl-koder for your comment.
